tracing-flame/src/lib.rs

```diff
@@ -0,0 +1,277 @@
+//! A Tracing [Layer][`FlameLayer`] for generating a folded stack trace for generating flamegraphs
+//! and flamecharts with [`inferno`]
+//!
+//! # Overview
+//!
+//! [`tracing`] is a framework for instrumenting Rust programs to collect
+//! scoped, structured, and async-aware diagnostics. This crate provides helpers
+//! for consuming tracing instrumentation for visualization as a
+//! flamegraph/flamechart which can be useful for finding performance
+//! bottlenecks.
+//!
+//! ## Usage
+//!
+//! This crate is meant to be used as a two step process. First, you capture a
+//! textual representation of the spans that are entered and exited with the
+//! [`FlameLayer`] then you feed these into `inferno-flamegraph` to generate the
+//! flamegraph/flamechart image.
+//!
+//! ## Layer Setup
+//!
+//! ```rust
+//! fn setup_global_subscriber() {
+//!     use tracing_flame::FlameLayer;
+//!     use tracing_subscriber::{registry::Registry, prelude::*};
+//!
+//!     let subscriber = Registry::default()
+//!         .with(FlameLayer::write_to_file("./tracing2.folded").unwrap());
+//!
+//!     tracing::subscriber::set_global_default(subscriber).expect("Could not set global default");
+//! }
+//!
+//! // your code here ..
+//! ```
+//!
+//! Alternatively you can provide any type that implements `std::io::Write` to
+//! `FlameLayer::new`.
+//!
+//! ## Generating The Image
+//!
+//! To convert your text representation to a visual one first install inferno
+//!
+//! ```console
+//! cargo install inferno
+//! ```
+//!
+//! Then pass the file your error layer created into `inferno-flamegraph`
+//!
+//! ```console
+//! # flamegraph
+//! cat tracing.folded | inferno-flamegraph > tracing-flamegraph.svg
+//!
+//! #flamechart
+//! cat tracing.folded | inferno-flamegraph --flamechart > tracing-flamechart.svg
+//! ```
+//!
+//! ## Differences between `flamegraph`s and `flamechart`s
+//!
+//! By default `inferno-flamegraph` creates `flamegraph`s. Flamegraphs is that
+//! they collapse identical stack frames and sort them based on their names.
+//! This is great for multithreaded programs and long running programs where the
+//! same frames happen many times for short durations because it simplifies the
+//! graph and gives you a broad idea of overall time spent in each part of the
+//! stack.
```

Comment:
```suggestion
//! By default, `inferno-flamegraph` creates flamegraphs. Flamegraphs operate by
//! that collapsing identical stack frames and sorting them on the frame's names.
//! 
//! This behavior is great for multithreaded programs and long-running programs
//! where the same frames occur _many_ times, for short durations, because it reduces
//! noise in the graph and gives the reader a better idea of the 
//! overall time spent in each part of the application.
```